apisix/admin/resource.lua

```diff
@@ -31,49 +32,89 @@ local mt = {
 }
 
 
-function _M:check_conf(id, conf, need_id)
+local no_id_res = {
+    consumers = true,
+    plugin_metadata = true
+}
+
+
+local function split_typ_and_id(id, sub_path)
+    local uri_segs = core.utils.split_uri(sub_path)
+    local typ = id
+    local id = nil
+    if #uri_segs > 0 then
+        id = uri_segs[1]
+    end
+    return typ, id
+end
+
+
+function _M:check_conf(id, conf, need_id, typ)
+    if self.name == "secrets" then
+        id = typ .. "/" .. id
+    end
     -- check if missing configurations
     if not conf then
         return nil, {error_msg = "missing configurations"}
     end
 
     -- check id if need id
-    id = id or conf.id
-    if need_id and not id then
-        return nil, {error_msg = "missing ".. self.kind .. " id"}
-    end
+    if not no_id_res[self.name] then
+        id = id or conf.id
+        if need_id and not id then
+            return nil, {error_msg = "missing ".. self.kind .. " id"}
+        end
 
-    if not need_id and id then
-        return nil, {error_msg = "wrong ".. self.kind .. " id, do not need it"}
-    end
+        if not need_id and id then
+            return nil, {error_msg = "wrong ".. self.kind .. " id, do not need it"}
+        end
 
-    if need_id and conf.id and tostring(conf.id) ~= tostring(id) then
-        return nil, {error_msg = "wrong ".. self.kind .. " id"}
-    end
+        if need_id and conf.id and tostring(conf.id) ~= tostring(id) then
+            return nil, {error_msg = "wrong ".. self.kind .. " id"}
+        end
 
-    conf.id = id
+        conf.id = id
+    end
 
-    core.log.info("schema: ", core.json.delay_encode(self.schema))
     core.log.info("conf  : ", core.json.delay_encode(conf))
 
     -- check the resource own rules
-    local ok, err = self.checker(id, conf, need_id, self.schema)
+    if self.schema then
+        typ = self.schema
+    else
```

Comment:
It would be better to check if the name == "secrets". As this is the only case when schema is nil, and we already have many `name == "secrets"` checks.